pkg/apiserver/authentication/oauth/groupmapper.go

```diff
@@ -0,0 +1,92 @@
+package oauth
+
+import (
+	"fmt"
+
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	userapi "github.com/openshift/api/user/v1"
+	"github.com/openshift/origin/pkg/cmd/server/apis/config"
+	"github.com/openshift/origin/pkg/user/apis/user/validation"
+	usercache "github.com/openshift/origin/pkg/user/cache"
+)
+
+func NewGroupMapper(groupCache *usercache.GroupCache, identityProviders []config.IdentityProvider) GroupMapper {
+	idpGroupsPrefix := map[string]string{
+		"": "", // prevents having to special case tokens that have no groups (and thus no provider name / prefix)
+	}
+	for _, identityProvider := range identityProviders {
+		providerName := identityProvider.Name
+		if identityProvider.LocalGroups {
+			idpGroupsPrefix[providerName] = "" // no prefix since we are told not to "scope" these groups
+		} else {
+			// build a prefix based on the name of the IDP (IDP names are unique)
+			// we use a / so that we are guaranteed to never conflict with builtin
+			// groups like system:masters and group+user API objects
+			idpGroupsPrefix[providerName] = providerName + "/"
```

Comment:
prefixing

Reply:
if a group or user name is used in paths, then using "/" will cause issues, we should probably choose a different separator

Reply:
No sure what user has to do with this.

That aside, where would the group be used in a path?  We basically need something that can never conflict with an existing group name.

Anything with a `/` or `%` will never conflict.

Anything with a `:` will not conflict as long we make sure it does not have something like `system:` in the front.  So it could be `${BASE} + ":" + idpName + ":"`.  The question then would be deciding on a `BASE` that makes sense.  `system:openshift:idp`?  Seems weird to use `system:openshift:..` but `idp` by itself seems weird too.

Reply:
are you really afraid someone will call their idp "system" ?
if that's the case let's just use "::" as the separator ?

Reply:
JFR: usernames are often used as path components, I can imagine stuff that may use group names too (for example in URLs), it makes no sense to make our life hard with needing escaping for each group name later when we can find better alternatives now.

Reply:
> are you really afraid someone will call their idp "system" ?

Yeah.

>  if that's the case let's just use "::" as the separator ?

That can also work.

Reply:
I honestly wouldn't be concerned about someone calling their IDP "system", or we could simply forbid that and be done, but then we would never be able to use another custom prefix if we need it.
So either we go with idp_<idpname>:<groupname> or <idpname>::<groupname>
I guess idp:<idpname>:<groupname> is also another option, or group:<idpname>:<groupname>

I prefer any of these forms to using a '/' separator

Reply:
> we could simply forbid that

We cannot do that retroactively though.  The IDP name is part of the identity objects - it is unchangeable once set.

---

So if we consider that we have `system:` and `system:openshift:` today, I am leaning towards:

1. Reserve `idp:` but do not use it for anything right now (in case we need something else IDP specific later)
2. Reserve `idp:groups:` and use it to prefix so you get: `idp:groups:<idp_name>:<group_name>`

Of course this is all arbitrary but it avoids `/` and gives us enough flexibility for the future.  Hopefully it is also obvious to anyone looking at the final string that we are talking about groups from a specific IDP.

Reply:
sounds reasonable
now id:groups: or idp:group: ?

Reply:
I used `groups` everywhere in this PR - i.e. `GroupsPrefix`, `GroupsHeaders`, `GroupsClaims`, etc.